rfcs/37-dynamic_query.md

```diff
@@ -0,0 +1,168 @@
+reliance on Rusts type system limits the types of queries which can be performed in `bevy_ecs`, as queries 
+must be defined at compile time. The `dynamic_query` RFC provides an alternative interface for performing 
+queries using `ComponentId`s at runtime, as the next step towards exposing the full power of bevy to scripting 
+and other types of dynamic plugin loading.
+
+## User-facing explanation
+
+As a bevy developer with a single vision who wants the best performance availabe, dynamic queries are strictly 
+inferior to their statically typed counterparts. However, developers who want to add support for scripting and mods 
+need a method of interacting with the Entity-Component-System at runtime, which is a niche filled by the 
+`DynamicQuery` type.
+
+Using static typing, you would perform a query like this:
+
+```rust
+let query_state = world.query::<(Entity, &Velocity, &mut Position, Option<&Layer>), Without<Frozen>>();
+```
+
+While the dynamic alternative looks like the following:
+
+```rust
+let dynamic_query = DynamicQuery::new()
+                        .entity()
+                        .component(velocity_component_id)
+                        .mut_component(position_component_id)
+                        .optional_component(layer_component_id)
+                        .without(frozen_component_id)
+                        .build();
+
+let query_state = world.query_dynamic(&query);
+```
+
+A static query returns the various components it has retrieved in a tuple. This isn't an option for a Dynamic Query
+however, since the number of components being returned isn't known at compile time. Instead, a Dynamic Query returns 
+a `DynamicQueryEntity` struct, which is both indexable and iterable to access the individual `DynamicItem`s it
+contains.
+
+A `DynamicItem` is an enum, which can represent the different types returned in a query result. Here are its
+variants:
+```rust
+pub enum DynamicItem {
+    Entity(Entity),
+    Component(DynamicComponentReference),
+    MutableComponent(DynamicMutComponentReference),
+    ComponentNotPresent
+}
+```
+The two inner types, `DynamicComponentReference` and `DynamicMutComponentReference` provide safe access to 
+`&`/`&mut` and change tracking functionality while also enabling users to have access to the underlying pointer for 
+more advanced usage.
+
+```rust
+let dynamic_query = DynamicQuery::new()
+                        .entity()
+                        .component(velocity_component_id)
+                        .mut_component(position_component_id)
+                        .without(frozen_component_id)
+                        .build();
+
+for result in world.query_dynamic(&query).iter_mut() {
+    info!("Updating the position of Entity {}", result[0].unwrap_entity().id())
+    result[2].unwrap_mut_component().downcast_mut::<Position>().unwrap() += result[1].unwrap_component().downcast_mut::<Velocity>.unwrap();
+}
+```
+
```

Comment:
IMO this really needs to show a more representative use case after the existing "basic usage" example.

Right now, all of the functionality demonstrated is simply less safe and slower than the existing tools :p